test/qa/lib/firewall.py

```diff
@@ -95,157 +95,220 @@
-                    f"-A INPUT -i {interface} -p {protocol} -m {protocol} --dport {port} -m comment --comment nordvpn -j ACCEPT",
-                    f"-A INPUT -i {interface} -p {protocol} -m {protocol} --sport {port} -m comment --comment nordvpn -j ACCEPT"
-                ])
-        result += \
-        [
-            f"-A INPUT -i {interface} -m connmark --mark 0xe1f1 -m comment --comment nordvpn -j ACCEPT",
-            f"-A INPUT -i {interface} -m comment --comment nordvpn -j DROP",
-        ]
-
-        for port in ports:
-            result.extend([
-                f"-A OUTPUT -o {interface} -p {protocol} -m {protocol} --dport {port} -m comment --comment nordvpn -j ACCEPT",
-                f"-A OUTPUT -o {interface} -p {protocol} -m {protocol} --sport {port} -m comment --comment nordvpn -j ACCEPT"
-            ])
+    # Connected
+    if ports == "" and protocol == "" and subnets == "":
+        return _get_rules_connected_to_vpn_server(interface)
 
-        result += \
-        [
-            f"-A OUTPUT -o {interface} -m mark --mark 0xe1f1 -m comment --comment nordvpn -j CONNMARK --save-mark --nfmask 0xffffffff --ctmask 0xffffffff",
-            f"-A OUTPUT -o {interface} -m connmark --mark 0xe1f1 -m comment --comment nordvpn -j ACCEPT",
-            f"-A OUTPUT -o {interface} -m comment --comment nordvpn -j DROP"
-        ]
-        return result
+    # Connected & Subnet(s) and Port(s) allowlisted
+    if subnets != "" and ports != "" and protocol == "":
+        return _get_rules_allowlist_subnet_and_port_on(interface, subnets, ports)
 
+    # Connected & Subnet(s) allowlisted
+    if subnets != "" and ports == "" and protocol == "":
+        return _get_rules_allowlist_subnet_on(interface, subnets)
 
-def is_active(ports: list[str]="", protocol="", subnets: list[str]="") -> bool:
-    """ returns True when all expected rules are found in iptables in matching order """
-    print(sh.ip.route())
+    # Connected & Port(s) allowlisted
+    if ports != "" and protocol == "":
+        return _get_rules_allowlist_port_on(interface, ports)
 
-    default_route_interface = sh.ip.route.show("default").split(None)[4]
+    # Connected & Port(s) allowlisted for specific protocols
+    if ports != "":
+        return _get_rules_allowlist_port_on_protocols(interface, ports, protocol)
 
-    try:
-        server_ip = settings.get_server_ip()
-    except IndexError:
-        server_ip = ""
 
-    print("Default gateway:", default_route_interface, "; Server IP:", server_ip)
+def is_active(ports: list[str]="", protocol="", subnets: list[str]="") -> bool:
+    """ returns True when all expected rules are found in iptables, in matching order """
+    print(sh.ip.route())
 
-    expected_rules = _get_firewall_rules(
-        daemon.is_killswitch_on(), server_ip, default_route_interface, ports, protocol, subnets
-    )
-    print("Expected rules:\n")
+    expected_rules = _get_firewall_rules(ports, protocol, subnets)
+    print("\nExpected rules:")
```

Comment:
Are prints even visible when running the tests? I thought that pytest captures the output.

Reply:
Same as in previous code, prints are only visible, if test fails :smile: